node/pkg/watchers/ibc/watcher.go

```diff
@@ -0,0 +1,734 @@
+		obsvReqC     <-chan *gossipv1.ObservationRequest
+	}
+)
+
+// NewWatcher creates a new IBC contract watcher
+func NewWatcher(
+	wsUrl string,
+	lcdUrl string,
+	contractAddress string,
+	chainConfig ChainConfig,
+) *Watcher {
+	return &Watcher{
+		wsUrl:           wsUrl,
+		lcdUrl:          lcdUrl,
+		contractAddress: contractAddress,
+		chainConfig:     chainConfig,
+		connectionIdMap: make(map[string]string),
+		connectionMap:   make(map[string]*connectionEntry),
+		chainMap:        make(map[vaa.ChainID]*connectionEntry),
+	}
+}
+
+// clientRequest is used to subscribe for events from the contract.
+type clientRequest struct {
+	JSONRPC string `json:"jsonrpc"`
+	// A String containing the name of the method to be invoked.
+	Method string `json:"method"`
+	// Object to pass as request parameter to the method.
+	Params [1]string `json:"params"`
+	// The request id. This can be of any type. It is used to match the
+	// response with the request that it is replying to.
+	ID uint64 `json:"id"`
+}
+
+// ibcReceivePublishEvent represents the log message received from the IBC receiver contract.
+type ibcReceivePublishEvent struct {
+	ChannelID string
+	Msg       *common.MessagePublication
+}
+
+// Run is the runnable for monitoring the IBC contract on wormchain.
+func (w *Watcher) Run(ctx context.Context) error {
+	w.logger = supervisor.Logger(ctx)
+
+	errC := make(chan error)
+	defer close(errC)
+
+	// Query the contract for the chain ID to IBC connection ID mapping.
+	connectionIdMap, err := w.queryConnectionIdMap()
+	if err != nil {
+		return fmt.Errorf("failed to query for connection ID map: %w", err)
+	}
+
+	// Build our internal data structures based on the config passed in.
+	if len(w.connectionMap) == 0 {
+		features := ""
+		for _, chainToMonitor := range w.chainConfig {
+			_, exists := w.chainMap[chainToMonitor.ChainID]
+			if exists {
+				return fmt.Errorf("detected duplicate chainID: %v", chainToMonitor.ChainID)
+			}
+
+			connID, exists := connectionIdMap[chainToMonitor.ChainID]
+			if !exists {
+				return fmt.Errorf("there is no IBC connection ID defined for chainID %v", chainToMonitor.ChainID)
+			}
+
+			ce := &connectionEntry{
+				connectionID: connID,
+				chainID:      chainToMonitor.ChainID,
+				chainName:    vaa.ChainID(chainToMonitor.ChainID).String(),
+				readiness:    common.MustConvertChainIdToReadinessSyncing(chainToMonitor.ChainID),
+				msgC:         chainToMonitor.MsgC,
+				obsvReqC:     chainToMonitor.ObsvReqC,
+			}
+
+			w.logger.Info("will monitor chain over IBC", zap.String("chain", ce.chainName), zap.String("IBC connection", ce.connectionID))
+			w.connectionMap[ce.connectionID] = ce
+			w.chainMap[ce.chainID] = ce
+
+			if features == "" {
+				features = "ibc:"
+			} else {
+				features += "|"
+			}
+			features += ce.chainID.String()
+
+			p2p.DefaultRegistry.SetNetworkStats(ce.chainID, &gossipv1.Heartbeat_Network{ContractAddress: w.contractAddress})
+		}
+
+		Features = features
```

Comment:
this is somewhat racy because anyone accessing Features (and performing a copy) before Run is executed will see an empty string.